modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts

```diff
@@ -0,0 +1,149 @@
+import { createHash } from 'crypto';
+import { bitLength, randBits, isProbablyPrime } from 'bigint-crypto-utils';
+import { gcd, modInv, modPow } from 'bigint-mod-arith';
+import { bigIntFromBufferBE, bigIntToBufferBE } from '../../util';
+import { DeserializedPaillierBlumProof } from './types';
+
+// Security parameter.
+const m = 80;
+
+/**
+ * Generate psuedo-random quadratic residue for (N, w, i).
+ * @param N - the prime number to verify is a product of two large primes.
+ * @param w - a random number with the same bitLength as N, that satisfies the Jacobi of w is -1 wrt N.
+ * @returns {bigint[]} - set of challenges for N
+ */
+function generateY(N, w): bigint[] {
+  const NBuf = bigIntToBufferBE(N);
+  const wBuf = bigIntToBufferBE(w, NBuf.length);
+  let counter = 0;
+  return Array(m)
+    .fill(null)
+    .map((_) => {
+      while (true) {
+        let h = bigIntFromBufferBE(
+          // TypeScript doesn't like us using `outputLength` in the transform options,
+          // but it is required for shake256.
+          createHash('shake256', { outputLength: Math.floor((bitLength(N) + 7) / 8) })
+            .update(Buffer.from([counter++]))
+            .update('$')
+            .update(NBuf)
+            .update('$')
+            .update(wBuf)
+            .update('$')
+            .digest()
+        );
+        h = (h * h) % N;
+        if (gcd(h, N) === BigInt(1)) {
+          return h;
```

Comment:
The sampling of y here deviates from https://eprint.iacr.org/2020/492.pdf section 4.3 and the non-interactive version at https://www.zkdocs.com/docs/zkdocs/zero-knowledge-protocols/product-primes/paillier_blum_modulus/#non-interactive-protocol

Here it seems to be sampling y so that it will be the same as y' by implicitly making a = 0, b = 0 (and thus not including a, b in the proof).  This will make only certain co-primes of N be picked as the challenge y, violation generateY() being a random oracle.

Reply:
Also, for y' (and therefor y here) to be suitable so that its 4th root is well defined, it needs to be a quadratic residue modulo N.  Here by making y = h^2 and y (mod N) = 1, y is indeed a quadratic residue N but only for the cases where modulo is 1.  Other quadratic residues are missed where x^2 === y (mod N) is other values (e.g. 4, 9, etc.).  This further limits the sampling domain of the challenge value.

Reply:
Could refer to Fireblocks' implementation https://github.com/fireblocks/mpc-lib/blob/main/src/common/crypto/paillier/paillier_zkp.c#L648

Basically, due to p, q === 3 mod 4 and jacobi(w, N) === -1, it's guaranteed one of (0, 0), (0, 1), (1, 0), (1,1) for (a, b) will make y' a quadratic residue modulo N.

Reply:
Done